### PR TITLE
feat: change avatar management

### LINKED ISSUE #702: Troubleshooting: Upload of avatars fails
Hi,

after uploading an avatar to a contact the image won't appear.

- Monica runs self hosted
- I cannot find any images under storage/app/public
- There is no folder 'avatars'. Should one be there?
- The folder is writable
- A Symlink from public is set
- The "FollowSymlinks" Setting is active

Is it possible to log the request? It seems like it cannot save the image, altough the browser is uploading it.

Thanks for your help!

### LINKED ISSUE #703: Troubleshooting: Displaying of Avatars
I am hosting Monica in IIS, I finally got the security settings right so that PHP could write the uploaded avatars to disk. When I upload an avatar it writes the image (plus the 110 and 174 resolutions) to Monica\storage\app\public\avatars. From the front end the image was broken with a 404. In one of the other issues I saw this command `php artisan storage:link` I noticed it was talking about a storage folder in public. After copying and pasting the avatars to Monica\public\storage\avatars the images were displayed.

I assume that `php artisan schedule:run` is meant to copy these images to the write folder? I havent seen this command actually do anything yet (other that return "No scheduled commands are ready to run.")

### LINKED ISSUE #2032: Dashboard avatars of inconsistant size and shape
Image avatars show up as an inconsistent size, depending on how many images v.s. initials are in the list of last consulted.

[A list heavy with images looks ALMOST normal](https://photos.app.goo.gl/bZAVN77AbdNd83X56)

[A list with only 4 images is barely useful](https://photos.app.goo.gl/pP1p4hKPUQ7u9Beh7)

[A list containing only 3 reduces the images to useless slivers.](https://photos.app.goo.gl/TUkFoXk371tjvNW57)

Note I am using the hosted version on https://app.monicahq.com

Thanks for listening!

### LINKED ISSUE #2145: Mention from where avatar photo is
I created a new contact and "magically" it has a avatar photo. I believe the photo is from Gravatar would be great to mention it to the user.

### PR DESCRIPTION
**Pull request by djaiss**
This PR completely changes how avatars are managed.

- We now have different avatar choices:
    - the default avatar that is now an image
    - adorable avatar
    - gravatar
    - photo that you upload
    - one of the photos on the contact page
- Default avatars are now an image that is created based on the initials of the contact. This image is stored on the filesystem. I've decided to use an image inside of HTML, so we can actually link to it, and therefore, the API will always return an image regardless of the avatar you use. Before, clients had to figure out which avatar they would use if the user didn't upload an avatar image.
- There is a new avatar that is pulled from http://avatars.adorable.io. This avatar is unique per contact (and uses a random UUID to identify it so no information is shared)
- The choice of using a gravatar in the Change avatar screen only appears if an email on the profile contact page is recognized by Gravatar.
- Any photo can be set as avatar
- If you delete a photo that is used as avatar, the avatar defaults back to the default one for this contact
- Rewrote how contacts are created and updated, using services

This close #2032 
This close #2145
This close https://github.com/monicahq/monica/issues/715
This close #702
This close #703

Note: ⚠️after migrating data, you need to run `php artisan monica:moveavatarstophotosdirectory` to move all the current avatars to the new Photo directory, as avatars are now actually Photo objects.

There are several things to do in another PR once we merge this PR:
* remove `gravatar_url`, `avatar_location`, `avatar_file_name`, `avatar_external_url`, `has_avatar` from the DB, which will be not used anymore. We first need to migrate all avatars to remove them completely, so it's better to do it in another PR.
* rewrite contact creation and update API calls. they still use the old methods but if I do it in this PR, the PR will be way too big.
* when destroying a contact, delete uploaded documents and photos. Again, for the sake of having the shorter PR, I didn't do it here.

**Tasks remaining to be done**

- [x] Clean up some tests files which use the old avatar
- [x] Take care of backward compatibility to the mobile app
- [x] Make sure the seeder works with the new avatar
- [x] Update the Contact resource with the new avatar properties
- [x] Also, fix the API
- [x] Migrate existing photo avatars to a Photo object
- [x] Replace all avatars on the dashboard
- [x] Replace avatars on the contact list
- [x] Replace avatars on the journal
- [x] Avatar in search
- [x] Make sure exporting account data as SQL is still working.
- [x] Make sure your changes do not break importing data with `vCard` and `.csv` files.
    - especially the use of the method `updateGravatar()`
    - all `gravatar_url`, `avatar_location`, `avatar_file_name`, `avatar_external_url`, `has_avatar` properties
- [x] Make sure account reset and deletion still work.
- [x] [CHANGELOG](https://github.com/monicahq/monica/blob/master/CHANGELOG) entry added, if necessary, under `UNRELEASED`.
- [ ] If it's relevant and worth mentioning, create a changelog entry for this change. The changelog entry will appear inside the UI for all users to see. To know if your change is worth the creation of a changelog entry, [read the documentation](https://github.com/monicahq/monica/blob/master/docs/administrators/tips.md#when-is-it-relevant-to-create-a-changelog-entry).

### CODE REVIEW

**Comment by asbiin** [comment]
Wow!

File: app/Helpers/AvatarHelper.php [hunk 1]
```diff
@@ -30,4 +30,14 @@ public static function get(Contact $contact, $size)
...
     }
+
+    /**
+     * Generate a new UUID to grab a new adorable avatar.
+     *
+     * @return string
+     */
+    public static function generateAdorableUUID()
+    {
+        return str_random(32);
```
[the remark this answers is not in the document]
Comment (djaiss) [reply app/Helpers/AvatarHelper.php]: You are right. Thanks!

**Comment by djaiss** [comment]
It's taking me much longer than I anticipated. Avatars are actually used everywhere and I also want to maintain compatibility with the existing mobile app, but I'm not sure if I should.

**Comment by djaiss** [comment]
I've worked a lot on this tonight.

In order to clean how avatars are handled, I had to clean the code in the contact object when we store a contact and when we update it. One thing leading to another… I rewrote how we manage create/update/delete a contact by creating services for those actions.

**Comment by djaiss** [comment]
@asbiin a lot has been done in this PR.

The thing I really want a feedback on would be this file: https://github.com/monicahq/monica/pull/2112/files#diff-2cea8fbc9429a2c0b51d0f7d76cec7ef

(MoveAvatarsToPhotosDirectory.php)

**Comment by djaiss** [comment]
@asbiin do you understand why this fails?

**Comment by scott-joe** [comment]
Does this address the variable avatar height issue?
**Without extra Stylus added CSS**
Screen Shot 2019-11-20 at 16 47 21
**With extra Stylus added CSS**
Screen Shot 2019-11-20 at 16 47 42

```css
div.avatar-search.avatar-initials {
    line-height: 31px;
    padding-bottom: 6px;
    border-radius: 2px;
    width: 43px;
}

**Comment by scott-joe** [comment]
Oh, it's been merged already. Guess not?

**Comment by asbiin** [comment]
@scott-joe the problem is still there, but as there are now 1 avatar file per contact, you can't see it. Your solution is nice, I will fix it as soon as possible.

**Comment by scott-joe** [comment]
> @scott-joe the problem is still there, but as there are now 1 avatar file per contact, you can't see it. Your solution is nice, I will fix it as soon as possible.

Cool! Happy to help :) I'd love to get the thing running locally to help. Just gotta find some free time :P

**Comment by SyntaxBlitz** [comment]
It took me some digging to find this issue so I could migrate my avatars after updating monica (from v2.14). Is there a way to get this built into the main `php artisan monica:update --force` migration script? I was surprised to have to do it manually.

**Pull request merged**